experimental/LieAlgebras/src/RootSystem.jl

```diff
@@ -0,0 +1,489 @@
+
+mutable struct RootSystem
+  cartan_matrix::ZZMatrix # (generalized) Cartan matrix
+  #fw::QQMatrix # fundamental weights as linear combination of simple roots
+  positive_roots::Any #::Vector{RootSpaceElem} (cyclic reference)
+  weyl_group::Any     #::WeylGroup (cyclic reference)
+  type::Vector{Tuple{Symbol,Int}}
+
+  function RootSystem(mat::ZZMatrix)
+    roots, refl = positive_roots_and_reflections(mat)
+    finite = count(refl .== 0) == nrows(mat)
+
+    R = new(mat)
+    R.positive_roots = map(r -> RootSpaceElem(R, r), roots)
+    R.weyl_group = WeylGroup(finite, refl, R)
+
+    return R
+  end
+end
+
+@doc raw"""
+    root_system(cartan_matrix::ZZMatrix) -> RootSystem
+    root_system(cartan_matrix::Matrix{Int}) -> RootSystem
+
+Constructs the root system defined by the Cartan matrix.
+"""
+function root_system(cartan_matrix::ZZMatrix)
+  return RootSystem(cartan_matrix)
+end
+
+function root_system(cartan_matrix::Matrix{<:Integer})
+  return RootSystem(matrix(ZZ, cartan_matrix))
+end
+
+@doc raw"""
+    root_system(fam::Symbol, rk::Int) -> RootSystem
+
+Constructs the root system of the given type. See `cartan_matrix(fam::Symbol, rk::Int)` for allowed combinations.
+"""
+function root_system(fam::Symbol, rk::Int)
+  cartan = cartan_matrix(fam, rk)
+  R = root_system(cartan)
+  R.type = [(fam, rk)]
+  return R
+end
+
+function root_system(types::Tuple{Symbol,Int}...)
+  cartan = cartan_matrix(types...)
+  R = root_system(cartan)
+  R.type = types
+  return R
+end
+
+function Base.show(io::IO, R::RootSystem)
+  println(io, "Root system defined by Cartan matrix")
+  show(io, cartan_matrix(R))
+end
+
+@doc raw"""
+    cartan_matrix(R::RootSystem) -> ZZMatrix
+
+Returns the Cartan matrix defining `R`.
+"""
+function cartan_matrix(R::RootSystem)
+  return R.cartan_matrix
+end
+
+#function basis(R::RootSystem)
+#  rk = rank(R)
+#  it = 1:rk
+#  return [RootSpaceElem(R, matrix(QQ, 1, rk, i .== it)) for i in 1:rk]
+#end
+
+function fundamental_weight(R::RootSystem, i::Int)
+  @req 1 <= i <= rank(R) "invalid index"
+  return WeightLatticeElem(R, matrix(ZZ, rank(R), 1, i .== 1:rank(R)))
+end
+
+@doc raw"""
+    fundamental_weights(R::RootSystem) -> Vector{WeightLatticeElem}
+
+Returns the fundamental weights corresponding to the simple roots of `R`.
+"""
+function fundamental_weights(R::RootSystem)
+  return [fundamental_weight(R, i) for i in 1:rank(R)]
+end
+
+function has_root_system_type(R::RootSystem)
+  return isdefined(R, :type)
+end
+
+function is_simple(R::RootSystem)
+  if has_root_system_type(R)
+    return length(root_system_type(R)) == 1
+  end
+  error("Not implemented") # TODO: implement is_simple
+end
+
+function negative_root(R::RootSystem, i::Int)
+  return -R.positive_roots[i]::RootSpaceElem
+end
+
+function negative_roots(R::RootSystem)
+  return [-r for r in positive_roots(R)]
+end
+
+function num_positive_roots(R::RootSystem)
+  return length(R.positive_roots)
```

Comment:
This suffers from type instability, can be solved via
```suggestion
  return length(positive_roots(R))
```
Same for most other places accessing `R.positive_roots` instead of `positive_roots(R)`